File: 290 Word Pattern/Java/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean wordPattern(String pattern, String s) {
        int len = pattern.length();
        String[] words = s.split(" ");

        if (words.length != len) {
            return false;
        }

        Map<Character, String> pToS = new HashMap<>();
        Map<String, Character> sToP = new HashMap<>();
        for (int i = 0; i < len; i++) {
            char c = pattern.charAt(i);
            String w = words[i];

            if (!pToS.containsKey(c) && !sToP.containsKey(w)) {
                pToS.put(c, w);
                sToP.put(w, c);
            }
            if (pToS.containsKey(c) && !w.equals(pToS.get(c)) ) {
                return false;
            }
            if (sToP.containsKey(w) && c != sToP.get(w)) {
                return false;
            }
        }

        return true;
    }
}

File: 290 Word Pattern/Python/Solution.py
from typing import Dict, List, Self


class Solution:
    def wordPattern(self: Self, pattern: str, s: str) -> bool:
        l: int = len(pattern)
        sList: List[str] = s.split(' ')
        if len(sList) != l:
            return False

        pToS: Dict[str, str] = {}
        sToP: Dict[str, str] = {}

        for i in range(l):
            char: str = pattern[i]
            word: str = sList[i]
            
            if char not in pToS and word not in sToP:
                pToS[char] = word
                sToP[word] = char
            if char in pToS and word != pToS.get(char):
                return False
            if word in sToP and char != sToP.get(word):
                return False
        return True